src/server/rdb_save.cc

```diff
@@ -189,16 +190,75 @@ class ZstdCompressImpl {
   base::PODArray<uint8_t> compr_buf_;
 };
 
-io::Bytes ZstdCompressImpl::Compress(io::Bytes str) {
-  size_t buf_size = ZSTD_compressBound(str.size());
+io::Result<io::Bytes> ZstdCompressor::Compress(io::Bytes data) {
+  size_t buf_size = ZSTD_compressBound(data.size());
   if (compr_buf_.capacity() < buf_size) {
     compr_buf_.reserve(buf_size);
   }
   size_t compressed_size = ZSTD_compressCCtx(cctx_, compr_buf_.data(), compr_buf_.capacity(),
-                                             str.data(), str.size(), compression_level_);
+                                             data.data(), data.size(), compression_level_);
 
+  if (ZSTD_isError(compressed_size)) {
+    return make_unexpected(error_code{int(compressed_size), generic_category()});
+  }
   compressed_size_total_ += compressed_size;
-  uncompressed_size_total_ += str.size();
+  uncompressed_size_total_ += data.size();
+  return io::Bytes(compr_buf_.data(), compressed_size);
+}
+
+class Lz4Compressor {
+ public:
+  // create a compression context
+  Lz4Compressor() {
+    LZ4F_errorCode_t result = LZ4F_createCompressionContext(&ctx_, LZ4F_VERSION);
+    if (LZ4F_isError(result)) {
+      // TODO this can fail on memory allocation. What should we do in this case?
```

Comment:
@romange how do you suggest to handle memory allocation errors? should we check fail? 
should I set and fail state to the compressor in the case of fail and check it status when using the object?

Reply:
it happens once when we first use it, right?
in that case - check fail is fine

Reply:
ok